### PR TITLE
[REVIEW] use installed C++ RMM in python build

### PR DESCRIPTION
**Pull request by germasch**
This PR switches the build of the python rmm to use the installed C++ RMM.

For the most part, this is just a cleanup, since the installed header-only RMM is the same as the one in the source tree.

But combined with PR #580, it will address issues like #584, where there is no spdlog preinstalled on the system. Currently, cmake will pull in its own spdlog in this case, but that spdlog won't be available to build the python modules. #580 will install spdlog together with RMM in this case, and with this PR, the python rmm build will use it.

Also, on a side note, for all I can tell, right now the build of the python modules will not use the downloaded Thrust 1.10.0, either, but whatever Thrust will be found on the Python side (ie., probably the one from the CUDA toolkit).

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Review by harrism — APPROVED** [review]

**Comment by harrism** [comment]
ok to test

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by harrism** [comment]
@germasch just informing you since you are a prolific external contributor to RMM now. We are nearing the end of the 0.16 dev cycle: Schedule here. We are entering the burndown phase tomorrow. In a nutshell, burndown means "no new PRs". If you think something should get into 0.16, please open a PR today, unless it's a significant change, in which case it should probably wait until 0.17. Tomorrow branch-0.17 will be created so you can target it with new PRs. See https://docs.rapids.ai/releases/process/

**Comment by kkraus14** [comment]
For what it's worth, touching the `setup.py` in this way is an area that doesn't have coverage in CI so I'd prefer to be extra cautious and not have this target 0.16 right before release.

Assigning to 0.17 for this reason and will update the branch once created.

**Comment by germasch** [comment]
I agree that this should be checked over carefully. But I also think the changes are small enough that it is possible for an experienced python person to review this and be pretty confident that nothing unexpected will happen (the maybe most complex part, getting `INSTALL_PREFIX` into `setup.py` follows how it's already done for `CUDA_HOME`).

I don't really care myself (I'm only using the C++ side of RMM), but it's worth mentioning that it fixes an actual issue (#584), which I had reproduced here. I didn't put a "fixes xxx" github comment to that effect since it's only true after my other cmake PR is merged.

**Comment by germasch** [comment]
Or maybe I can't get it past the style checker, anyway :(

When I ran `black` (version 20.8b1) locally, it reformatted `setup.py` and `rmm.py`, the latter wasn't even touched by this PR. In any case, your CI black is still not happy with what my black did...

**Comment by germasch** [comment]
Got the formatting thing worked out. Since #580 is merged, fixes #584.

**Review by harrism — APPROVED** [review]

File: python/setup.py
Comment (kkraus14) [inline python/setup.py:60]: Not everyone uses the `build.sh` so we can't rely on the `INSTALL_PREFIX` being set when someone runs `python setup.py`

**Review by kkraus14 — CHANGES_REQUESTED** [review]
Haven't done an in depth review yet but needed to block merging

File: python/setup.py
Comment (germasch) [reply python/setup.py:60]: So the idea behind using the rmm "install" (even it may just be in a subdir of the build directory) is that the python modules are built against not only the rmm headers (which are the same as the ones in source dir, so this doesn't really make a difference), but also against other thirdparty dependencies that RMM pulled in via FetchContent, like spdlog or, eventually, Thrust.

This is, admittedly, never going to be perfect. In the case of pre-installed dependencies, there's no guarantee that the installed version of spdlog that cmake detects is going to be the same that ends up being used in the python build. Though it'll likely work in most standard situations, except for the case covered here, ie., cmake downloaded spdlog and the python build would have no idea about that without this PR.

I guess the best I can come up with to fix your concern is to handle the case of `INSTALL_PREFIX` not being set above and falling back to the previous hard-coded relative path. How's that?

Comment (kkraus14) [reply python/setup.py:60]: > This is, admittedly, never going to be perfect. In the case of pre-installed dependencies, there's no guarantee that the installed version of spdlog that cmake detects is going to be the same that ends up being used in the python build. Though it'll likely work in most standard situations, except for the case covered here, ie., cmake downloaded spdlog and the python build would have no idea about that without this PR.

It's on my TODO list to explore using [scikit-build](https://scikit-build.readthedocs.io/en/latest/) for handling building the Python extensions. This should allow us to reuse the existing CMake machinery we've already built and make the native code dependency handling much cleaner from the Python side.

> I guess the best I can come up with to fix your concern is to handle the case of INSTALL_PREFIX not being set above and falling back to the previous hard-coded relative path. How's that?

That makes sense to me!

Comment (germasch) [reply python/setup.py]: Okay, put the fallback back in.

I didn't really know about scikit-build, but at least from a quick look at their docs, it does seem appealing ;)

File: python/setup.py
Comment (kkraus14) [inline python/setup.py:60]: Can we include `../build/include` in here as well just to be safe?

Comment (germasch) [reply python/setup.py:60]: I can if you really want me to, but this directory doesn't even get created anymore. I believe it was used for `cnmem.h`, which is gone.

Comment (kkraus14) [reply python/setup.py:60]: Apologies, I thought that they were copied into a `../build/include` directory for some reason. This should be good to go, thanks!

**Review by kkraus14 — APPROVED** [review]

**Pull request merged**
